include/bluetooth/l2cap.h

```diff
@@ -191,10 +180,23 @@ struct bt_l2cap_le_chan {
 	struct k_work			tx_work;
 	/** Segment SDU packet from upper layer */
 	struct net_buf			*_sdu;
-	uint16_t				_sdu_len;
+	uint16_t			_sdu_len;
 
 	struct k_work			rx_work;
 	struct k_fifo			rx_queue;
+
+#if defined(CONFIG_BT_L2CAP_DYNAMIC_CHANNEL)
```

Comment:
My understanding from @jhedberg was that `struct bt_l2cap_chan` represents a fixed channel, and `struct bt_l2cap_le_chan` represents a dynamic channel. If this is true, then we shouldn't need this guard here.

Reply:
I don't modify it like that, and it doesn't seem necessary, and this structure should be a common part

Reply:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/att.c#L2738

I don't want to involve too much code. Maybe I need to see @jhedberg 's opinion. If the existing method is unreasonable, I don't need to follow up your comments. Just close it directly.

Reply:
Yes, the point was that `bt_l2cap_chan` should be a common "parent" struct, but also sufficient in itself for fixed channels. I.e. even if you've enabled dynamic channel support in Kconfig, your fixed channel instances shouldn't consume more memory than necessary for a fixed channel. You can also go and have a more complex hierarchy with a common struct, and then a derived fixed channel struct and a derived dynamic channel struct, however from what I can see there would be no difference between the common struct and the fixed channel struct in that case, so you can skip this additional complexity.

Reply:
@LingaoM what do you mean?

Reply:
> `struct bt_l2cap_chan` represents a fixed channel, and `struct bt_l2cap_le_chan` represents a dynamic channel

This is false. `bt_l2cap_chan_send` always down-casts `bt_l2cap_chan` to `bt_l2cap_le_chan` if the connection is BT LE. `bt_l2cap_chan` is the abstraction over BT LE and BT BR channels. For BT LE, both fixed and dynamic channels must be a `bt_l2cap_le_chan` object.

Reply:
@alwa-nordic I don't think that's correct. If you look at `subsys/bluetooth/host/l2cap.c` all places that use `bt_l2cap_le_chan` are related to dynamic channels. At least I didn't find any code there that would exclude `bt_l2cap_chan` from the LE bearer. This also makes a lot of sense, since when the stack got originally created fixed channels and LE was all that was supported, and we had `bt_l2cap_chan` already back then. It's true that ATT uses `bt_l2cap_le_chan` but I suspect that's so that the same abstraction can be used regardless of what kind of transport the ATT channel resides on (you'll see a lot of places in the code referring to `chan.chan` since it needs to use `bt_l2cap_chan` instead of `bt_l2cap_le_chan`.

Reply:
@alwa-nordic I see your point about `bt_l2cap_chan_send()` though. I need to look a bit more into this, but as far as I can remember, the intended design is still that `bt_l2cap_chan` is sufficient for fixed channels.

Reply:
Ok, so I didn't need to dig deep for the answer. The only scenario where `att.c` uses `bt_l2cap_chan_send()` is for dynamic channels with EATT. For the "normal" fixed channel ATT the code uses `bt_l2cap_send_cb()`, which doesn't even bother with the `bt_l2cap_chan` abstraction, but just passes the CID as a direct parameter.

Reply:
@alwa-nordic there's a lot of things to be cleaned up in the code. I hope you're looking with equal attention at the points which contradict your premise as those which seem to support it. I think what you've found is one of the areas where further cleanup/refactoring may be needed. It seems @Vudentz (who I would *really* hope would jump in to comment here) has taken some shortcuts/simplifications in the current design by reusing variables that were originally only intended for dynamic channels.

Reply:
To get this discussion to a more constructive track, I'd propose to move to defining how we'd want the L2CAP design to be, both from a logical structure perspective, as well as from a resource usage perspective (which is what this PR is about to begin with).

The state and properties that you need to track for a fixed channel are in practice a subset of those for dynamic channels (whether LE or BR/EDR), so it makes IMO sense to have that as a parent struct from which dynamic channel structs are derived. The code already does this to a large extent, but it seems the current stack-internal fixed channels (ATT & SMP) have been implemented in a way that reuse state from the `bt_l2cap_le_chan` struct. The extent of reuse seems to be super minimal however, essentially only limited to tracking the RX & TX CID values (which are one and the same for fixed channels).

Reply:
@LingaoM Consider this thread resolved. The ifdef is necessary for now for optimizing memory use. In the future we want to make it so `bt_l2cap_le_chan` will never be instantiated if dynamic channels are disabled. Then the ifdef will be superfluous.